planemo/galaxy/workflows.py

```diff
@@ -364,21 +375,20 @@ def _job_outputs_template_from_invocation(invocation_id, galaxy_url, galaxy_api_
         outputs[label] = {"path": f"test-data/{label}.{ext}"}
     for label, output in invocation["output_collections"].items():
         collection = user_gi.dataset_collections.show_dataset_collection(output["id"])
-        if ":" not in collection["collection_type"]:
-            user_gi.datasets.download_dataset(
-                collection["elements"][0]["object"]["id"],
-                use_default_filename=False,
-                file_path=f"test-data/{label}.{collection['elements'][0]['object'].get('file_ext', 'txt')}",
-            )
-            outputs[label] = {
-                "element_tests": {  # only check the first element
-                    collection["elements"][0]["element_identifier"]: {
-                        "path": f"test-data/{label}.{collection['elements'][0]['object']['file_ext']}"
-                    }
-                }
-            }
-        else:
-            outputs[label] = {"element_tests": "nested_collection_todo"}
+        outputs[label] = test_dict = {}
+        for collection_type in collection["collection_type"].split(":"):
+            assert collection_type in ("list", "paired", "record"), f"{collection_type} not valid"
+            element = collection["elements"][0]  # only check the first element
+            element_identifier = element["element_identifier"]
+            test_dict = test_dict[element_identifier] = {}
+        file_path = f"test-data/{label}.{element['object'].get('file_ext', 'txt')}"
+        user_gi.datasets.download_dataset(
+            element["object"]["id"],
+            use_default_filename=False,
+            file_path=file_path,
+        )
+        test_dict["element_tests"] = {element_identifier: {"path": file_path}}
+        outputs[label] = {"element_tests": test_dict}
```

Comment:
To be honest, stuff like `test_dict = test_dict[element_identifier] = {}` seems like black magic to me (it sparked some fascinated discussion in our team chat), but I don't see how this produces a nested output - won't it just create
```
outputs:
  label:
    element_tests:
      element_identifier:
        path: file_path
```
like in https://github.com/galaxyproject/iwc/pull/106?

Reply:
You can write this as
```
test_dict["element_identifier"] = {}
test_dict = test_dict["element_identifier"]
```
dicts are mutable in python, so you're building up a nested dict with each level of the collection nesting.

Reply:
Ha, glad you're confused by this, too :wink: This would result in a simple empty dict.
The real equivalent would be:
```
test_dict = {}
test_dict["element_identifier"] = test_dict
```
which at least to me makes the recursion much more obvious.

Reply:
Y'all, this is really standard python, and that's not the equivalent @wm75:
```

In [1]: outputs = {}

In [2]: outputs['some_output'] = test_dict = {}

In [3]: test_dict["element_identifier"] = {}

In [4]: test_dict = test_dict["element_identifier"]

In [5]: test_dict['nested'] = 'yes'

In [6]: outputs
Out[6]: {'some_output': {'element_identifier': {'nested': 'yes'}}}
```

Reply:
That's the point I guess. Your `[3]` and `[4]` are *not* the same as `test_dict = test_dict[element_identifier] = {}`.

Reply:
I should have added some tests from the start, added that now in https://github.com/galaxyproject/planemo/pull/1242/commits/e03f75465b8ac761d7df2198dacffc56530320cb and I reworked the whole logic to just be one recursion for generating input and output definitions.